README.md
#  Dip Authenticator -  Android OTP Authenticator

[![Build Status](https://travis-ci.org/Dip Authenticator/Dip Authenticator.svg?branch=master)](https://travis-ci.org/Dip Authenticator/Dip Authenticator)
[![Current release](https://img.shields.io/github/release/Dip Authenticator/Dip Authenticator/all.svg)](https://github.com/Dip Authenticator/Dip Authenticator/releases/download/v0.6.0-beta3/Dip Authenticator_v0.6.0-beta3.apk)
[![Crowdin](https://d322cqt584bo4o.cloudfront.net/andotp/localized.svg)](https://crowdin.com/project/andotp)
[![Chat - Telegram](https://img.shields.io/badge/chat-Telegram-blue.svg)](https://t.me/Dip Authenticator)
[![Chat - Matrix](https://img.shields.io/badge/chat-Matrix-blue.svg)](https://matrix.to/#/#Dip Authenticator:matrix.shadowice.org)

![Dip Authenticator](./assets/logo.png)

Dip Authenticator is a two-factor authentication App for Android 4.4+.

It implements Time-based One-time Passwords (TOTP) and HMAC-Based One-Time Passwords (HOTP).
Simply scan the QR code and login with the generated 6-digit code.

This project started out as a fork of the great [OTP Authenticator](https://github.com/0xbb/otp-authenticator) app written by [Bruno Bierbaumer](https://github.com/0xbb),
which has sadly been inactive since 2015. By now almost every aspect of the app has been changed/re-written so the fork status of the Github repository got
detached upon [user request](https://github.com/Dip Authenticator/Dip Authenticator/issues/145). But all credit for the original version and for starting this project still goes to Bruno!

## Help wanted:
I currently don't have that much time to spend developing Dip Authenticator, so any contributions are always welcome.
Don't worry, I will still continue to develop Dip Authenticator it will just slow down from the incredible speed I had going in the beginning.

## Features:

 * Free and Open-Source
 * Requires minimal permissions
   - Camera access for QR code scanning
   - Storage access for import and export of the database
 * Encrypted storage with two backends:
   - Android KeyStore
   - Password / PIN
 * Multiple backup options:
   - Plain-text
   - Password-protected
   - OpenPGP-encrypted
 * Sleek minimalistic Material Design with three different themes:
   - Light
   - Dark
   - Black (for OLED screens)
 * Great Usability
 * Compatible with Google Authenticator
 * Supported algorithms:
   - TOTP (Time-based One-time Passwords) as specified in [RFC 6238](https://tools.ietf.org/html/rfc6238)
   - HOTP (HMAC-based One-time Passwords) as specified in [RFC 4226](https://tools.ietf.org/html/rfc4226)

## Backups:

To keep your account information as secure as possible Dip Authenticator only stores it in
encrypted data files. A part of the encryption key used for that is stored in the
Android KeyStore system. The advantage of this approach is that the key is kept
separate from the apps data and, as a bonus, can be backed by hardware cryptography
(if your device supports this).

However, due to that separation, backups with 3rd-party apps like Titanium Backup can not
be used with Dip Authenticator. Such apps only backup the encrypted data files and not the encryption
key, which renders them useless.

**Please only use the internal backup functions provided by Dip Authenticator to backup your accounts!**
**Everything else WILL result in data loss.**

### Opening the backups on your PC:

 * [OpenPGP](http://openpgp.org/): OpenPGP can be used to easily decrypt the OpenPGP-encrypted backups on your PC.
 * [Dip Authenticator-decrypt](https://github.com/asmw/Dip Authenticator-decrypt): A python script written by @asmw to decrypt password-protected backups on your PC (needs more testing).

### Automatic backups:

 * BroadcastReceivers: AndOTP supports a number of broadcasts to perform automated backups, eg. via Tasker. These will get saved to the defined backup directory. **These only work when KeyStore is used as the encryption mechanism**
   - **org.shadowice.flocke.andotp.broadcast.PLAIN_TEXT_BACKUP**: Perform a plain text backup. **WARNING**: This will save your 2FA tokens onto the disk in an unencrypted manner!
   - **org.shadowice.flocke.andotp.broadcast.ENCRYPTED_BACKUP**: Perform an encrypted backup of your 2FA database using the selected password in settings.

## Migration:

Check out [this](https://github.com/Dip Authenticator/Dip Authenticator/wiki/Migration) wiki page to learn about the different ways to migrate to Dip Authenticator from other 2FA apps.

## Downloads:

[<img height=80 alt="Get it on Google Play" src="https://play.google.com/intl/en_us/badges/images/generic/en-play-badge.png" />](https://play.google.com/store/apps/details?id=org.shadowice.flocke.andotp)
[<img height=80 alt="Get it on F-Droid" src="https://f-droid.org/badge/get-it-on.png" />](https://f-droid.org/packages/org.shadowice.flocke.andotp/)
[<img height=80 alt="Get it on GitHub" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/badges/get-it-on-github.png" />](https://github.com/Dip Authenticator/Dip Authenticator/releases)

**Warning**: All three versions (Google Play, F-Droid and the APKs) are not compatible (not signed by the same key)!
You will have to uninstall one to install the other, which will delete all your data.
So make sure you have a **current backup** before switching!

## Contribute:

 * **Translation**: If you want to help translate Dip Authenticator into your language head over to the [Crowdin project](https://crowdin.com/project/andotp).
 * **Bug reports and feature requests**: You can report bugs and request features in the [Issue tracker](https://github.com/Dip Authenticator/Dip Authenticator/issues) on GitHub.
 * **Requesting thumbnails**: If you are missing a thumbnail you can request it by editing the [wiki](https://github.com/Dip Authenticator/Dip Authenticator/wiki/Thumbnails#thumbnail-requests)
 * **Discussion and support**: 
   - [XDA thread](https://forum.xda-developers.com/android/apps-games/app-andotp-android-otp-authenticator-t3636993) (please keep off-topic to a minimum)
   - Telegram group [@Dip Authenticator](https://t.me/Dip Authenticator) (if you just want important updates you can mute the group so you only get notified about pinned messages)
   - Matrix room [#Dip Authenticator:matrix.shadowice.org](https://matrix.to/#/#Dip Authenticator:matrix.shadowice.org) (in case you prefere Matrix over Telegram)

#### Developers:

 * [Jakob Nixdorf](https://github.com/flocke) (Telegram: [@flocke000](https://t.me/flocke000), Reddit: [/u/flocke000](https://www.reddit.com/user/flocke000))
 * [Richy HBM](https://github.com/RichyHBM)

#### Contributors (in order of appearance):

 * [Carlos Melero](https://github.com/carmebar) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits/master?author=carmebar))
 * [SuperVirus](https://github.com/SuperVirus) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits/master?author=SuperVirus))
 * [DanielWeigl](https://github.com/DanielWeigl) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits/master?author=DanielWeigl))
 * [Matthias Bertram](https://github.com/mbertram) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits?author=mbertram))
 * [Björn Richter](https://github.com/x3rAx) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits?author=x3rAx))
 * [Emmanuel Nwokoma](https://github.com/enwokoma) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits?author=enwokoma))
 * [magnus anderssen](https://github.com/callmemagnus) ([view contributions](https://github.com/Dip Authenticator/Dip Authenticator/commits?author=callmemagnus))

#### Translators:

&nbsp; | Language          | Translators
------ | ----------------- | -----------
🇵🇱   | Polish (pl-rPL)   | [Daniel Pustuła](https://github.com/9Cube-dpustula), [bla_ke](https://crowdin.com/profile/bla_ke), [DamianR](https://crowdin.com/profile/DamianR)
:es:   | Spanish (es-rES)  | [Carlos Melero](https://crowdin.com/profile/carmebar)
:de:   | German (de-rDE)   | [SuperVirus](https://crowdin.com/profile/SuperVirus), [Jan](https://crowdin.com/profile/Dagefoerde), [TMHBOFH](https://crowdin.com/profile/TMHBOFH), [Grammost](https://crowdin.com/profile/Grammost), [Andreas Kleinert](https://crowdin.com/profile/AndyKl)
:fr:   | French (fr-rFR)   | [Johan Fleury](https://github.com/johanfleury), [David Sferruzza](https://crowdin.com/profile/dsferruzza), [primokorn](https://crowdin.com/profile/primokorn), [Poussinou](https://crowdin.com/profile/Poussinou), [Patrick Decat](https://crowdin.com/profile/pdecat)
🇳🇱   | Dutch (nl-rNL)    | Toon, [rain2reign](https://crowdin.com/profile/rain2reign), [thinkwell](https://crowdin.com/profile/thinkwell), [cpu20](https://crowdin.com/profile/cpu20)
&nbsp; | Galician (gl-rES) | [Triskel](https://crowdin.com/profile/triskel), [Xosé M.](https://crowdin.com/profile/XoseM)
:ru:   | Russian (ru-rRU)  | [Victor Nidens](https://crowdin.com/profile/vnidens), [Ilia Drogaitsev](https://crowdin.com/profile/waytoroot), [Dmitry](https://crowdin.com/profile/SaintDI), [Анастасия Дрогайцева](https://crowdin.com/profile/khaleesi16)
🇨🇿   | Czech (cs-rCZ)    | [Picard0147](https://crowdin.com/profile/Picard0147)
:cn: | Chinese Simplified (zh-rCN) | [Cp0204](https://crowdin.com/profile/Cp0204)
&nbsp; | Catalan (ca-rES) | [isard](https://crowdin.com/profile/isard)

## Screenshots:
#### Light theme:

[<img width=200 alt="Main Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity.png)
[<img width=200 alt="Settings Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity.png)
[<img width=200 alt="Backup Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity.png)

#### Dark theme:

[<img width=200 alt="Main Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity_dark.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity_dark.png)
[<img width=200 alt="Settings Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity_dark.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity_dark.png)
[<img width=200 alt="Backup Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity_dark.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity_dark.png)

#### Black theme:

[<img width=200 alt="Main Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity_black.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/main_activity_black.png)
[<img width=200 alt="Settings Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity_black.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/settings_activity_black.png)
[<img width=200 alt="Backup Activity" src="https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity_black.png">](https://raw.githubusercontent.com/flocke/Dip Authenticator/master/assets/screenshots/backup_activity_black.png)

## Acknowledgments:
#### Open-source components used:

 * [AboutLibraries](https://github.com/mikepenz/AboutLibraries)
 * [Apache Commons Codec](https://commons.apache.org/proper/commons-codec/)
 * [Expandable Layout](https://github.com/AAkira/ExpandableLayout)
 * [Floating Action Button Speed Dial](https://github.com/leinardi/FloatingActionButtonSpeedDial)
 * [material-intro](https://github.com/heinrichreimer/material-intro)
 * [MaterialProgressBar](https://github.com/DreaminginCodeZH/MaterialProgressBar)
 * [OpenPGP API library](https://github.com/open-keychain/openpgp-api)
 * [VNTNumberPickerPreference](https://github.com/vanniktech/VNTNumberPickerPreference)
 * [ZXing Android Embedded](https://github.com/journeyapps/zxing-android-embedded)

#### Code examples used:

 * [Android-ItemTouchHelper-Demo](https://github.com/iPaulPro/Android-ItemTouchHelper-Demo/tree/master/app/src/main/java/co/paulburke/android/itemtouchhelperdemo/helper)
 * [Code Parts from Google's Android Samples](https://android.googlesource.com/platform/development/+/master/samples/Vault/src/com/example/android/vault)
 * [LetterBitmap](http://stackoverflow.com/questions/23122088/colored-boxed-with-letters-a-la-gmail)
 * [DimensionConverter](https://stackoverflow.com/questions/8343971/how-to-parse-a-dimension-string-and-convert-it-to-a-dimension-value)

#### Previously used open-source components:

 * [FABsMenu](https://github.com/jahirfiquitiva/FABsMenu)
 * [LicensesDialog](https://github.com/PSDev/LicensesDialog)

#### Previously used code examples:

 * [FloatingActionMenuAndroid](https://github.com/pmahsky/FloatingActionMenuAndroid)

## License:
```
Copyright (C) 2017-2018 Jakob Nixdorf <flocke@shadowice.org>
Copyright (C) 2015 Bruno Bierbaumer

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in the
Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
OR OTHER DEALINGS IN THE SOFTWARE.
```


CHANGELOG.md
# Changelog

#### v0.6.0

 * New feature: **HOTP support**
 * New feature: Settings item to activate Broadcast backups
 * New feature: Re-locking of the app on screen off is now optional (Issue #28)
 * New feature: Allow PGP backups with only a public key (Issue #31)
 * New feature: Show individual timeout bars on the cards for non-default periods
 * New feature: App shortcuts to add new entries (Issue #185)
 * Bug fix: Avoid crash on empty PIN/Password an API 23 (Issue #159, PR #160 by magnus anderssen)
 * Bug fix: Honor the system accessibility settings for the font size (Issue #71, PR #192 by @mbertram)
 * Internal: Replace custon FAB menu with Floating Action Button Speed Dial library (Issue #155 and #186)
 * Style/UI: Use AboutLibraries instead of LicenseDialog and rework the About section (Issue #155)
 * Style/UI: Show a disclaimer about the included thumbnails in the About screen
 * Update: ZXing Android Embedded (3.6.0), Constraint Layout (1.1.2) and all support libraries (27.1.1)
 * F-Droid: Add the feature graphic and some screenshots (PR #117 by @jaller94)
 * Thumbnails: lots of new thumbnails (see the wiki)

#### v0.5.1 (Google Play only)

 * Like v0.5.0 but with Password-based encryption offered by default during setup

#### v0.5.0

 * New feature: **Intro screen when staring the app for the first time to setup encryption and authentication**
 * New feature: **Broadcast receivers to trigger backups from Tasker** (PR #115)
 * New feature: **Add support for using Android Backup** (Issue #109, PR #111)
 * New feature: Optionally append date to backups (PR #124)
 * New feature: Check if entries are valid when entering manually (Issue #135, PR #136 by Björn Richter)
 * New feature: Offer different options when using the tag selection (Issue #133, PR #134)
 * New feature: Show a warning before changing the encryption
 * Bug fix: Fix crash when saving an empty label (Issue #138, PR #139 by Björn Richter)
 * Bug fix: Fix visibility of thumbnails in dark themes (Issue #88, PR #90)
 * Bug fix: Don't require credentials again after screen rotation (Issue #152)
 * Thumbnails: new thumbnails (see the wiki)

#### v0.4.0

 * New feature: **Password-based encryption** (a big thanks to all the testers)
 * New feature: Enforce a minimum password / PIN length (Issue #107)
 * New feature: Add an additional unlock button to the authentication (Issue #87)
 * New feature: The thumbnail toggle is now in the size selector (Issue #98, PR #102)
 * New feature: Split the tokens into blocks (Issue #83, PR #83 by DanielWeigl)
 * New feature: Account name is now shown in the removal confirmation (Issue #84)
 * New feature: Advanced options are now hidden in the manual entry dialog (Issue #85)
 * New special feature: Clear the KeyStore (use with caution)
 * Bug fix: Change the format used to store and set the language (Issue #112)
 * Bug fix: Add some extra padding the the RecyclerView (Issue #95)
 * Bug fix: Remove gradients from vector thumbnails (Issue #103, PR #97)
 * Thumbnails: a lot of new thumbnails (check the wiki for details)
 * Translation: Catalan (ca-rES) thanks to isard

#### v0.3.1

 * Move: the Github repository was moved from flocke/Dip Authenticator to Dip Authenticator/Dip Authenticator for better organization of collaborators
 * New feature: assign (predefined) images to entries (Issue #14, PR #75, again thanks to @RichyHBM for the implementation)
 * New feature: sort labels locale-sensitive (PR #74 by carmebar)
 * New feature: re-hide the revealed entries after a configurable timeout (Issue #77)
 * New feature: add sorting by last usage (Issue #67)
 * New feature: improved error messages during the import of backups
 * New feature: make the replace switch default to false (Issue #80)
 * New special feature: disable Special features again
 * New special feature: enable screenshots in the main Activity
 * Bug fix: use sp for font sizes (to make them scalable)
 * Bug fix: disable the save button in a manual entry until label and secret are not empty (Issue #82)
 * Style/UI: better description of the replace switch
 * Update: Android SDK 27 (Issue #76)
 * Update: Android Gradle plugin 3.0.1
 * Translation: Chinese Simplified (zh-rCN) thanks to Cp0204

#### v0.3.0

 * New feature: tagging support (Issue #37, PR #64, big thanks to @RichyHBM for actually implementing this)
 * New feature: settings option to scroll overlong labels instead of just truncating them
 * New feature: option to append entries during import instead of just replacing everything
 * New feature: in-app language switcher (Issue #53)
 * Bug fix: convert secrets to upper case when importing from JSON (Issue #55)
 * Bug fix: some layout fixes for certain translations (Issue #58)
 * Style/UI: new adaptive icon for Android 8+ (Issue #65)
 * Style/UI: remove card elevation
 * Update: Android Studio 3
 * Update: Gradle 4.1 / Android Gradle Plugin 3.0

#### v0.2.8

 * New feature: store authentication credentials hashed (Issue #49)
 * New feature: store backup password encrypted (Issue #49)
 * New feature: set a static backup dir to disable the file selector (Issue #52)
 * New feature: special features (see wiki)
 * New special feature: SteamGuard tokens (Issue #38)
 * Style/UI: black theme (Issue #47)
 * Bug fix: keep authentication settings when receiving a Panic Trigger (Issue #50)
 * Bug fix: progress bar animation with default duration scale
 * Translation: Czech (cs-rCZ) thanks to Picard0147

#### v0.2.7

 * New feature: require authentication again after screen lock (Issue #28)
 * New feature: make response to Panic Trigger configurable (Issue #35)
 * Bug fix: prevent adding duplicate entries (Issue #41)
 * Update: Android SDK 26 (Oreo)
 * Update: Apache Commons Codec 1.10
 * Code: lot of internal changes (mostly due to the Android 26 update)
 * Translation: French (fr-rFR) thanks to Johan Fleury
 * Translation: Durch (nl-rNL) thanks to T-v-Gerwen and rain2reign
 * Translation: Galician (gl-rES) thanks to Triskel

app/src/main/java/org/shadowice/flocke/andotp/Utilities/Constants.java
/*
 * Copyright (C) 2017-2018 Jakob Nixdorf
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.shadowice.flocke.andotp.Utilities;

import android.os.Environment;

import com.google.android.gms.wallet.WalletConstants;

import java.io.File;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class Constants {
    // Enums
 /*   public enum AuthMethod {
        NONE, PASSWORD, PIN, DEVICE
    }*/

    // Intents (Format: A0x with A = parent Activity, x = number of the intent)
    public final static int INTENT_MAIN_AUTHENTICATE = 100;
    public final static int INTENT_MAIN_SETTINGS = 101;
    public final static int INTENT_MAIN_BACKUP = 102;
    public final static int INTENT_MAIN_INTRO = 103;
    public final static int INTENT_BACKUP_OPEN_DOCUMENT_PLAIN = 200;
    public final static int INTENT_BACKUP_SAVE_DOCUMENT_PLAIN = 201;
    public final static int INTENT_BACKUP_OPEN_DOCUMENT_CRYPT = 202;
    public final static int INTENT_BACKUP_SAVE_DOCUMENT_CRYPT = 203;
    public final static int INTENT_BACKUP_OPEN_DOCUMENT_PGP = 204;
    public final static int INTENT_BACKUP_SAVE_DOCUMENT_PGP = 205;
    public final static int INTENT_BACKUP_ENCRYPT_PGP = 206;
    public final static int INTENT_BACKUP_DECRYPT_PGP = 207;
    public final static int INTENT_BACKUP_SAVE_CLOUD_CRYPT = 208;
    public final static int INTENT_BACKUP_OPEN_CLOUD_CRYPT = 209;
    public static final int INTENT_SETTINGS_AUTHENTICATE = 300;
    // Permission requests (Format: A1x with A = parent Activity, x = number of the request)
    public final static int PERMISSIONS_BACKUP_READ_IMPORT_PLAIN = 210;
    public final static int PERMISSIONS_BACKUP_WRITE_EXPORT_PLAIN = 211;
    public final static int PERMISSIONS_BACKUP_READ_IMPORT_CRYPT = 212;
    public final static int PERMISSIONS_BACKUP_WRITE_EXPORT_CRYPT = 213;
    public final static int PERMISSIONS_BACKUP_READ_IMPORT_PGP = 214;
    public final static int PERMISSIONS_BACKUP_WRITE_EXPORT_PGP = 215;
    // Intent extras
    public final static String EXTRA_AUTH_PASSWORD_KEY = "password_key";
    public final static String EXTRA_AUTH_NEW_ENCRYPTION = "new_encryption";
    public final static String EXTRA_AUTH_MESSAGE = "message";
    public final static String EXTRA_BACKUP_ENCRYPTION_KEY = "encryption_key";
    public final static String EXTRA_SETTINGS_ENCRYPTION_CHANGED = "encryption_changed";
    public final static String EXTRA_SETTINGS_ENCRYPTION_KEY = "encryption_key";
    // Authentication
    public final static int AUTH_MIN_PIN_LENGTH = 4;
    public final static int AUTH_MIN_PASSWORD_LENGTH = 6;
    // KeyStore
    public final static String KEYSTORE_ALIAS_PASSWORD = "password";
    public final static String KEYSTORE_ALIAS_WRAPPING = "settings";
    // Database files
    public final static String FILENAME_ENCRYPTED_KEY = "otp.key";
    public final static String FILENAME_DATABASE = "secrets.dat";
    public final static String FILENAME_DATABASE_BACKUP = "secrets.dat.bck";
    // Database files
    public final static String DIP_FILENAME_ENCRYPTED_KEY = "dipotp.key";
    public final static String DIP_FILENAME_DATABASE = "dipsecrets.dat";
    public final static String DIP_FILENAME_DATABASE_BACKUP = "dipsecrets.dat.bck";
    public final static String BACKUP_FOLDER_NAME = "Dip 2FA Backup";
    public final static String APP_FOLDER_NAME = "Dip Auth";
    public final static String AUTOBACKUP_FILE_NAME = "autobackup.json.aes";
    //    public final static String AUTOBACKUP_FILE_NAME = "autobackup_%s.json.aes";
    // Backup files
    public final static String BACKUP_FOLDER = Environment.getExternalStorageDirectory().getAbsolutePath() + File.separator + APP_FOLDER_NAME + File.separator + BACKUP_FOLDER_NAME;

    public final static String BACKUP_FILENAME_PLAIN = "otp_accounts.json";
    public final static String BACKUP_FILENAME_CRYPT = "otp_accounts.json.aes";
    public final static String BACKUP_FILENAME_PGP = "otp_accounts.json.gpg";

    public final static String BACKUP_FILENAME_PLAIN_FORMAT = "otp_accounts_%s.json";
    public final static String BACKUP_FILENAME_CRYPT_FORMAT = "otp_accounts_%s.json.aes";
    public final static String BACKUP_FILENAME_PGP_FORMAT = "otp_accounts_%s.json.gpg";

    public final static String BACKUP_MIMETYPE_PLAIN = "application/json";
    public final static String BACKUP_MIMETYPE_CRYPT = "binary/aes";
    public final static String BACKUP_MIMETYPE_PGP = "application/pgp-encrypted";
    public static final int PAYMENTS_ENVIRONMENT = WalletConstants.ENVIRONMENT_TEST;
    /**
     * The allowed networks to be requested from the API. If the user has cards from networks not
     * specified here in their account, these will not be offered for them to choose in the popup.
     *
     * @value #SUPPORTED_NETWORKS
     */
    public static final List<String> SUPPORTED_NETWORKS = Arrays.asList(
            "AMEX",
            "DISCOVER",
            "JCB",
            "MASTERCARD",
            "VISA");
    /**
     * The Google Pay API may return cards on file on Google.com (PAN_ONLY) and/or a device token on
     * an Android device authenticated with a 3-D Secure cryptogram (CRYPTOGRAM_3DS).
     *
     * @value #SUPPORTED_METHODS
     */
    public static final List<String> SUPPORTED_METHODS =
            Arrays.asList(
                    "PAN_ONLY",
                    "CRYPTOGRAM_3DS");
    /**
     * Required by the API, but not visible to the user.
     *
     * @value #COUNTRY_CODE Your local country
     */
    public static final String COUNTRY_CODE = "US";
    /**
     * Required by the API, but not visible to the user.
     *
     * @value #CURRENCY_CODE Your local currency
     */
    public static final String CURRENCY_CODE = "USD";
    /**
     * Supported countries for shipping (use ISO 3166-1 alpha-2 country codes). Relevant only when
     * requesting a shipping address.
     *
     * @value #SHIPPING_SUPPORTED_COUNTRIES
     */
    public static final List<String> SHIPPING_SUPPORTED_COUNTRIES = Arrays.asList("US", "GB");
    /**
     * The name of your payment processor/gateway. Please refer to their documentation for more
     * information.
     *
     * @value #PAYMENT_GATEWAY_TOKENIZATION_NAME
     */
    public static final String PAYMENT_GATEWAY_TOKENIZATION_NAME = "example";
    /**
     * Custom parameters required by the processor/gateway.
     * In many cases, your processor / gateway will only require a gatewayMerchantId.
     * Please refer to your processor's documentation for more information. The number of parameters
     * required and their names vary depending on the processor.
     *
     * @value #PAYMENT_GATEWAY_TOKENIZATION_PARAMETERS
     */
    public static final HashMap<String, String> PAYMENT_GATEWAY_TOKENIZATION_PARAMETERS =
            new HashMap<String, String>() {
                {
                    put("gateway", PAYMENT_GATEWAY_TOKENIZATION_NAME);
                    put("gatewayMerchantId", "exampleGatewayMerchantId");
                    // Your processor may require additional parameters.
                }
            };
    /**
     * Only used for {@code DIRECT} tokenization. Can be removed when using {@code PAYMENT_GATEWAY}
     * tokenization.
     *
     * @value #DIRECT_TOKENIZATION_PUBLIC_KEY
     */
//    public static final String DIRECT_TOKENIZATION_PUBLIC_KEY = "REPLACE_ME";
    public static final String DIRECT_TOKENIZATION_PUBLIC_KEY = "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEGnJ7Yo1sX9b4kr4Aa5uq58JRQfzD8bIJXw7WXaap\\/hVE+PnFxvjx4nVxt79SdRuUVeu++HZD0cGAv4IOznc96w==";
    /**
     * Parameters required for {@code DIRECT} tokenization.
     * Only used for {@code DIRECT} tokenization. Can be removed when using {@code PAYMENT_GATEWAY}
     * tokenization.
     *
     * @value #DIRECT_TOKENIZATION_PARAMETERS
     */
    public static final HashMap<String, String> DIRECT_TOKENIZATION_PARAMETERS =
            new HashMap<String, String>() {
                {
                    put("protocolVersion", "ECv2");
                    put("publicKey", DIRECT_TOKENIZATION_PUBLIC_KEY);
                }
            };
    // Encryption algorithms and definitions
    final static String ALGORITHM_SYMMETRIC = "AES/GCM/NoPadding";
    final static String ALGORITHM_ASYMMETRIC = "RSA/ECB/PKCS1Padding";
    final static int ENCRYPTION_KEY_LENGTH = 16;           // 128-bit encryption key (KeyStore-mode)
    final static int ENCRYPTION_IV_LENGTH = 12;
    final static int PBKDF2_MIN_ITERATIONS = 1000;
    final static int PBKDF2_MAX_ITERATIONS = 5000;
    final static int PBKDF2_DEFAULT_ITERATIONS = 1000;
    final static int PBKDF2_LENGTH = 256;      // 128-bit encryption key (Password-mode)
    final static int PBKDF2_SALT_LENGTH = 16;

    public enum AuthMethod {
        PIN
    }

    public enum EncryptionType {
        KEYSTORE, PASSWORD
    }

    public enum SortMode {
        UNSORTED, LABEL, LAST_USED
    }

    public enum BackupType {
        PLAIN_TEXT, ENCRYPTED, OPEN_PGP
    }

    public enum TagFunctionality {
        OR, AND, SINGLE
    }

    public enum NotificationChannel {
        BACKUP_FAILED, BACKUP_SUCCESS
    }
}
